lib/chef/provider/service/openbsd_rcctl.rb

```diff
@@ -0,0 +1,157 @@
+
+          requirements.assert(:start, :enable, :reload, :restart, :status) do |a|
+            a.assertion { service_exists? }
+            a.failure_message Chef::Exceptions::Service, "#{new_resource}: rcctl does not recognize this service."
+          end
+        end
+
+        # Enable service and optionally set any additional variables such as command line flags
+        # that are stored in the `parameters` hash.
+        #
+        # As of OpenBSD 5.7 the following variables can be set (rcctl(8)):
+        #
+        #   flags, timeout, user.
+        #
+        # This code does not strictly enforce these variables since rcctl(8) will
+        # exit non-zero if you give it something it does not recognize.
+        #
+        def action_enable
+          if current_resource.enabled && !parameters_need_update?
+            Chef::Log.debug("#{new_resource} already enabled - nothing to do")
+          else
+            converge_by("enable service #{new_resource}") do
+              update_params!
+              enable_service
+              Chef::Log.info("#{new_resource} enabled")
+            end
+          end
+          load_new_resource_state
+          new_resource.enabled(true)
+        end
+
+        def enable_service
+          shell_out!("rcctl enable #{new_resource.service_name}")
+        end
+
+        def disable_service
+          shell_out!("rcctl disable #{new_resource.service_name}")
+        end
+
+        private
+
+        # `rcctl get <service> status` returns:
+        # - 0: service is enabled
+        # - 1: service is not enabled
+        # - 2: service does not exist
+        def rcctl_status
+          shell_out!("rcctl get #{new_resource.service_name} status", env: nil, returns: [0, 1, 2]).status
+        end
+
+        def service_exists?
+          rcctl_status != 2
+        end
+
+        def is_enabled?
+          rcctl_status == 0
+        end
+
+        def rcctl_enable
+          shell_out!("rcctl enable #{new_resource.service_name}")
+        end
+
+        # ex:
+        #   $ rcctl get sshd
+        #   sshd_flags=foo
+        #   sshd_timeout=30
+        #   sshd_user=root
+        def determine_current_parameters!
+          current_resource.parameters Hash.new
+          shell_out!("rcctl get #{new_resource.service_name}", returns: [0, 1]).stdout.each_line do |line|
+            if line =~ /^#{Regexp.escape(new_resource.service_name)}_(.+?)=(.*)/
+              current_resource.parameters[Regexp.last_match(1)] = Regexp.last_match(2)
+            else
+              Chef::Log.warn("Problem parsing '#{line}' from 'rcctl get #{new_resource.service_name}'")
+              next
+            end
+          end
+          Chef::Log.debug("current params from #{new_resource}: #{current_resource.parameters}")
+        end
+
+        # check the current parameters against any newly specified parameters.
+        def parameters_need_update?
+          if new_resource.parameters && current_resource.parameters
```

Comment:
It looks like current_resource.parameters is always at least the empty hash, so testing if its set isn't necessary (and confused me for a second that you had a logic bug here).